Doc updates (#86)
amd/IRON: AGENTS.md

# AGENTS.md

This file provides guidance to AI coding agents when working with code in this repository.

## Overview

IRON is a close-to-metal Python API for AMD Ryzen™ AI NPUs (XDNA architecture). It provides language bindings around the MLIR-AIE dialect to enable fast and efficient execution on NPU hardware.

**Key Technologies:**

- **MLIR-AIE**: Dialect for programming AMD AI Engines (AIE) array architectures
- **XRT (Xilinx Runtime)**: Low-level runtime for interfacing with NPU hardware
- **Target Hardware**: AMD Ryzen AI NPUs (AIE2/AIE2P architectures - NPU1/NPU2)
- **Primary Datatype**: bfloat16

## Environment Setup

```bash
# 1. Source XRT (required for all operations)
source /opt/xilinx/xrt/setup.sh

# 2. Activate virtual environment
source ironenv/bin/activate

# 3. Install dependencies

> pip install -r requirements.txt

Reviewer: Would it make sense to separate out the environment _installation_ steps (which would also have to include `python3 -m venv ironenv`) from the environment _activation_ steps? Also I've had trouble with calling `pip` directly in some environments, for maximum compatibility I think `python3 -m pip` would help.

```

**Note:** XRT must be sourced before running any tests or operators.

### Build Directory

Compiled artifacts (`.xclbin`, `.bin`, `.o` files) are stored in `build/` directory by default. The build directory can be customized via `AIEContext(build_dir="path/to/build")`.

### Environment Variables

- `IRON_EXAMPLE_WEIGHTS_DIR`: Path to model weights for applications (default: `/srv`)

## Building and Testing

### Run All Operators (non-extensive tests)

```bash

> pytest iron/operators/ -m "not extensive"

Reviewer: For functional tests only I'd add an `--iterations 1` here.

```

### Run Extensive Test Suite

```bash
pytest iron/operators/
```

### Run Single Operator Test

```bash
pytest iron/operators/axpy/
# OR directly execute the test.py

> ./iron/operators/axpy/test.py

Reviewer 2: This does not run the test (it just does basic syntax check). It was true in the past, but wrong since tests were rewritten with pytest. README.md has the same mistake - it was not updated.

```

### Run Application Tests

```bash
pytest iron/applications/
```

### Run Specific Test Function

```bash

> pytest iron/operators/gemm/test.py::test_gemm_basic

Reviewer 2: Hallucinated, there is no `test_gemm_basic`

Author: There was one in my local changes 😄 I think it preferred that since I was asking a lot of questions about GEMMs.

- Used to parallelize work across multiple columns
- Format: `(tensor_shape, offset, dimensions, strides)`

**Runtime Sequence**: Host-side control flow

- `rt.fill()`: DMA data from host → NPU (shim → L2/L1)
- `rt.drain()`: DMA data from NPU → host
- `rt.start()`: Launch workers
- `rt.task_group()`: Coordinate parallel DMA operations

**Compilation Flow**:

```text
design.py (Python MLIR-AIE API)
    ↓
PythonGeneratedMLIRArtifact
    ↓
MLIR (.mlir file)
    ↓ (aie-opt + aie-translate via Peano toolchain)
xclbin (NPU binary) + insts.bin (instruction sequence)
```

**AIEContext**: Manages compilation and runtime state

- Default build directory: `build/` in current working directory
- Compilation rules: Defines pipeline from Python → MLIR → xclbin
- Device manager: Singleton for XRT resource sharing
- Use `AIEContext(build_dir="...", mlir_verbose=True)` for custom settings

**Device Manager**: Singleton that manages XRT resources

- Automatically initializes `pyxrt.device(0)`
- Caches contexts and kernels per xclbin path
- Shared across all operators to avoid resource conflicts

## Hardware Constraints

### NPU Architecture Limits

> - **NPU1 (AIE2)**: 4 rows × 1-4 columns (AMD Ryzen AI Phoenix/Hawk Point)

Reviewer: 4 rows x 4 accessible columns

Author: I don't think I understand the difference between the two.

Reviewer: In the other architecture you state 4 rows x 8 columns whereas here its 4 rows x 1-4 columns. But all Phoenix and Hawk Point devices have 4 rows x 4 columns as far as I know. Accessible you can leave off I just added that because technically there's 5 columns in hardware but the driver dropped support for using the first a while ago.

Author: I think it came to this conclusion because for the NPU1Col1 etc. objects. Maybe I can add that information instead of the 1-4

- **NPU2 (AIE2P)**: 4 rows × 8 columns (AMD Ryzen AI 300 Series "Strix Point", Ryzen AI 9 HX 370 "Strix Halo", Krackan)

### Tile and Dimension Constraints

Common operator parameters and their constraints:

> - `tile_size`: Typically 64, 128, 256, or 4096 (depends on operator and data type)

Reviewer: Not sure this is a hard requirement, perhaps this is more confusing than helpful?

Author: I found that I need to give it something for `tile_size` in designs outside IRON. Are the values too restrictive or just wrong?

Reviewer: They're not wrong, if it's helpful should be fine to leave them in.

- Minimum tile sizes depend on `emulate_bf16_mmul_with_bfp16` flag:
  - `True` (default): 8×8×8 minimum
  - `False`: 4×8×8 minimum
- Matrix dimensions must be multiples of `tile × num_rows/columns`
  - `M % (tile_m * 4) == 0`
  - `K % tile_k == 0`
  - `N % (tile_n * num_aie_columns) == 0`

**Element-wise ops** (add, mul, relu, gelu, etc.):

- `size % (num_aie_columns * tile_size) == 0`
- `size % tile_size == 0`

### Memory Hierarchy

- **L3**: Host memory (DDR)
- **L2**: Shared memory tiles (MemTiles in AIE-ML)
- **L1**: Per-core local memory (limited, ~32-64 KB per tile)

Data movement pattern: L3 → Shim DMA → L2 → L1 (tile local) → Compute

## Adding a New Operator

1. Create directory in `iron/operators/<operator_name>/`
2. Implement `op.py`:
   - Subclass `MLIROperator`
   - Implement `get_operator_name()`, `get_mlir_artifact()`, `get_kernel_artifacts()`, `get_arg_spec()`
   - Add validation for dimension constraints (assert statements)
   - Define tile sizes and column counts
3. Implement `design.py`:
   - Import from `aie.iron` (Program, Runtime, Worker, ObjectFifo, Kernel)
   - Define function that builds MLIR-AIE design

>    - Use `range_()` for loops (not Python `range`)

Reviewer: Would a "why" be useful here for the agents? Make it clear that this is metaprogramming that will translate directly to MLIR constructs such as scf.for

Author: I am not sure. These are instructions so that agents can understand the codebase. Understanding MLIR may be outside the scope of that (?)

   - Handle device-specific logic (NPU1 vs NPU2) if needed
4. Implement C++ kernel in `aie_kernels/<arch>/` if needed
   - Choose appropriate directory: `generic/`, `aie2/`, or `aie2p/`
   - Use AIE API for portable vectorization when possible
   - Add `event0()` and `event1()` for performance profiling
5. Implement `reference.py` with CPU reference
6. Implement `test.py` with pytest tests
   - Use `@pytest.mark.extensive` for slower/larger tests
   - Use `verify_buffer()` from `iron.common.test_utils`
7. Register operator in `iron/operators/__init__.py`

## Operator Fusion

> IRON supports fusing multiple operators into a single xclbin for improved performance:

Reviewer: ```suggestion
IRON supports fusing multiple operators into a single ELF file. This improves performance enabling a single runtime dispatch for a chain of operators. This works only with the "full ELF" flow, which uses ELF files at runtime. The ELF files take the place of `xclbin`s.
```

```python
from iron.common.fusion import FusedMLIROperator

# Define individual operators
gemm1 = AIEGEMM(...)
relu = AIERELU(...)
gemm2 = AIEGEMM(...)

# Create fused operator with runlist
# Intermediate buffers are automatically managed
fused_op = FusedMLIROperator(
    name="fused_gemm_relu_gemm",
    runlist=[
        (gemm1, "in", "temp1"),      # (operator, input_buffers, output_buffers)
        (relu, "temp1", "temp2"),
        (gemm2, "temp2", "out"),
    ],
    input_args={"in": size_in},
    output_args={"out": size_out},
    context=ctx
)
```

Benefits of fusion:

- Reduces host ↔ NPU data transfers

> - Shares compiled resources (kernels.a)

Reviewer: ```suggestion
```

> - Eliminates intermediate buffer round-trips

Reviewer: ```suggestion
- Runs a chain of operators using a single host-side dispatch (one CPU/host interrupt after fusion vs. one interrupt per operator without fusion)
```

## Common Patterns

### Multi-Column Parallelism

Distribute work across NPU columns using TensorAccessPattern:

```python
num_columns = 4
chunk = total_elements // num_columns

taps = [
    TensorAccessPattern(
        (1, total_elements),
        chunk * i,  # offset for column i
        [1, 1, 1, chunk],

>         [0, 0, 0, 1],

Reviewer: ```suggestion
        [1, 1, 1, chunk],  # sizes
        [0, 0, 0, 1],  # strides
```
